hypershift-operator/controllers/nodepool/haproxy.go

```diff
@@ -0,0 +1,381 @@
+	hyperv1 "github.com/openshift/hypershift/api/v1alpha1"
+	"github.com/openshift/hypershift/control-plane-operator/controllers/hostedcontrolplane/ignition"
+	"github.com/openshift/hypershift/hypershift-operator/controllers/hostedcluster"
+	"github.com/openshift/hypershift/hypershift-operator/controllers/manifests"
+	"github.com/openshift/hypershift/hypershift-operator/controllers/manifests/controlplaneoperator"
+	"github.com/openshift/hypershift/support/config"
+	"github.com/openshift/hypershift/support/globalconfig"
+	"github.com/openshift/hypershift/support/releaseinfo"
+	"github.com/openshift/hypershift/support/util"
+	mcfgv1 "github.com/openshift/hypershift/thirdparty/machineconfigoperator/pkg/apis/machineconfiguration.openshift.io/v1"
+	"github.com/vincent-petithory/dataurl"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/utils/pointer"
+	crclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	controlPlaneOperatorSkipsHAProxyConfigGenerationLabel = "io.openshift.hypershift.control-plane-operator-skips-haproxy"
+	haProxyRouterImageName                                = "haproxy-router"
+)
+
+func (r *NodePoolReconciler) reconcileHAProxyIgnitionConfig(ctx context.Context, releaseImage *releaseinfo.ReleaseImage, hcluster *hyperv1.HostedCluster, globalConfig globalconfig.GlobalConfig) error {
+	var pullSecret corev1.Secret
+	if err := r.Client.Get(ctx, types.NamespacedName{Namespace: hcluster.Namespace, Name: hcluster.Spec.PullSecret.Name}, &pullSecret); err != nil {
+		return fmt.Errorf("failed to get pull secret: %w", err)
+	}
+	pullSecretBytes, ok := pullSecret.Data[corev1.DockerConfigJsonKey]
+	if !ok {
+		return fmt.Errorf("expected %s key in pull secret", corev1.DockerConfigJsonKey)
+	}
+	controlPlaneOperatorImage, err := hostedcluster.GetControlPlaneOperatorImage(ctx, hcluster, r.ReleaseProvider, r.HypershiftOperatorImage, pullSecretBytes)
+	if err != nil {
+		return fmt.Errorf("failed to get controlPlaneOperatorImage: %w", err)
+	}
+	controlPlaneOperatorImageMetadata, err := r.ImageMetadataProvider.ImageMetadata(ctx, controlPlaneOperatorImage, pullSecretBytes)
+	if err != nil {
+		return fmt.Errorf("failed to look up image metadata for %s: %w", controlPlaneOperatorImage, err)
+	}
+	// Don't manage the configmap or we will race with the CPO
+	if _, cpoSkips := util.ImageLabels(controlPlaneOperatorImageMetadata)[controlPlaneOperatorSkipsHAProxyConfigGenerationLabel]; !cpoSkips {
+		return nil
+	}
+
+	hcpNS := manifests.HostedControlPlaneNamespace(hcluster.Namespace, hcluster.Name).Name
+	hcp := controlplaneoperator.HostedControlPlane(hcpNS, hcluster.Name)
+	if err := r.Client.Get(ctx, crclient.ObjectKeyFromObject(hcp), hcp); err != nil {
+		// Might not be created yet, we will get triggered once it is
```

Comment:
>IMHO setting such a boundary is pretty arbitrary, both the nodepool controller and the HCP are part of the project and their existence is an implementation detail.

Fair point. Aside from this particular PR just to provide you more context about my view: Generically I think current boundaries enable components hc/nodepool/hcp to be decoupled building blocks that can evolve independently with controlled friction. I think having these boundaries helps contain project complexity and enables the code base to grow and evolve in a sustainable way.
NodePool references HC in the API nodepool.spec.cluster so its consumption is organic and even though we try to keep the deps to a very minimum so machine management can evolve independently.
I.e imho these boundaries support a modularised design favouring cohesion and reducing coupling, which enables sustainability.